File: web/src/sessionStats.js
// The conversation header's statistics line.
//
// The numbers come from GET /api/chat/sessions/{id} (`stats`), which the server
// aggregates from what is stored — so they survive a reload and are the same for
// every client. This module only decides how they read: which fields are worth a
// segment, in what order, and what each one's tooltip says.
//
// It is a module of its own, and pure, because that decision is the part worth
// checking without a browser (see scripts/check-stats.mjs) — the panel is then a
// loop over the segments instead of five places that format a number.
import { formatCompact, formatCount, formatDuration } from './format.js'

/**
 * normalizeStats reads a `stats` payload into the fields this module renders.
 *
 * Every field is coerced to a non-negative integer: the header must not print
 * `NaN 次` because a server grew a field this build does not know yet.
 */
export function normalizeStats(raw) {
  const s = raw && typeof raw === 'object' ? raw : {}
  const n = (v) => {
    const value = Number(v)
    return Number.isFinite(value) && value > 0 ? Math.floor(value) : 0
  }
  return {
    messages: n(s.messages),
    turns: n(s.turns),
    llmCalls: n(s.llm_calls),
    llmDurationMs: n(s.llm_duration_ms),
    toolCalls: n(s.tool_calls),
    toolDurationMs: n(s.tool_duration_ms),
    promptTokens: n(s.prompt_tokens),
    completionTokens: n(s.completion_tokens),
    totalTokens: n(s.total_tokens),
  }
}

/**
 * statsSegments turns a normalized payload into the header's segments.
 *
 * A field with nothing to say is left out rather than rendered as a zero: a
 * conversation that has made no model call yet shows "轮 0" and stops there,
 * which is more informative than four zeros competing for the same line.
 *
 * `turns` is always present — it is the one number that is meaningful at 0, and
 * the line's anchor — and `messageCount` (the session's own count, passed in by
 * the header) sits directly behind it: together they say how much conversation
 * there is, before the segments that say what it cost. It is not read from the
 * payload because the session row carries it and it is the number the sidebar
 * counts; a header whose message count disagreed with the list would be worse
 * than one without it.
 */
export function statsSegments(raw, { messageCount } = {}) {
  const s = normalizeStats(raw)
  const segments = [
    {
      key: 'turns',
      text: `轮 ${formatCount(s.turns)}`,
      title: `共 ${formatCount(s.turns)} 轮：你发起的请求数`,
    },
  ]
  // A number is rendered even at 0 — like `turns`, "no messages yet" is a fact
  // worth the four characters. Absent (a caller that has no session) means the
  // segment is not this module's to invent.
  const messages = Number(messageCount)
  if (Number.isFinite(messages) && messages >= 0) {
    segments.push({
      key: 'messages',
      text: `共 ${formatCount(messages)} 条消息`,
      title: `共 ${formatCount(messages)} 条消息：你说的每一句和模型写下的每一段都算一条`,
    })
  }
  if (s.llmCalls > 0) {
    segments.push({
      key: 'llm',
      text: `模型 ${formatCount(s.llmCalls)} 次 · ${formatDuration(s.llmDurationMs)}`,
      title:
        `模型调用 ${formatCount(s.llmCalls)} 次，服务端计时合计 ${formatDuration(s.llmDurationMs)}` +
        '（每次 assistant 消息即一次调用；耗时是 provider 上报的单次耗时之和）',
    })
  }
  if (s.toolCalls > 0) {
    segments.push({
      key: 'tools',
      text: `工具 ${formatCount(s.toolCalls)} 次 · ${formatDuration(s.toolDurationMs)}`,
      title:
        `工具调用 ${formatCount(s.toolCalls)} 次，实测耗时合计 ${formatDuration(s.toolDurationMs)}` +
        '（来自每一轮存下来的过程记录，与每条回答下面那行「执行过程」同一份数据）',
    })
  }
  if (s.totalTokens > 0) {
    segments.push({
      key: 'tokens',
      // The total is the headline; the split is on hover, because it is the
      // detail that only matters once the total looks wrong.
      text: `${formatCompact(s.totalTokens)} tokens`,
      title:
        `prompt ${formatCount(s.promptTokens)} + completion ` +
        `${formatCount(s.completionTokens)} = ${formatCount(s.totalTokens)} tokens（provider 上报）`,
    })
  }
  return segments
}

/**
 * statsTitle is the whole line's tooltip: the segments' own titles, plus the one
 * caveat a reader would otherwise read as a bug.
 */
export function statsTitle(raw, options) {
  return [
    ...statsSegments(raw, options).map((segment) => segment.title),
    '两个耗时都不是这段对话的墙钟长度：排队、步与步之间的思考、以及你本人停顿的时间都不计在内。',
  ].join('\n')
}

File: internal/server/sessionstats.go
package server

import (
	"context"
	"encoding/json"

	"go.uber.org/zap"

	"github.com/huan/huan-agent/internal/chat"
	"github.com/huan/huan-agent/internal/store"
)

// sessionStats is what one conversation has cost and done so far, as the header
// of that conversation reports it.
//
// Every number is derived from what is stored, not from a counter kept in
// memory, so it survives a page reload, a model switch and a process restart,
// and it is the same answer for the console and for any other client reading the
// API. Three definitions are worth stating because a reader will otherwise guess
// them:
//
//   - LlmCalls counts assistant messages, because one assistant message is one
//     completed model call. A turn with ten steps contributes ten.
//   - ToolCalls counts the tool calls written down with the turn — read out of
//     the stored steps, which is the same list the transcript renders, so the
//     number in the header cannot disagree with the number of tool rows on
//     screen. A turn stored before steps were kept carries the same calls in the
//     flat ToolCalls column instead, and that is the fallback.
//   - ToolDurationMs is what the calls themselves reported, again from the steps
//     (the per-turn line under each answer shows the same sum). Older turns whose
//     steps recorded no timing fall back to the audit log's aggregate.
//
// Durations are what the provider reported for a call and what the tool itself
// measured, and neither is the wall-clock length of the conversation: they
// exclude queueing, thinking between steps, and the user's own time. The
// header's tooltip says so, since a total that looks too small is otherwise read
// as a bug.
type sessionStats struct {
	// Messages is every stored message, user and assistant and tool.
	Messages int `json:"messages"`
	// Turns counts the user messages, i.e. the requests that were made.
	Turns int `json:"turns"`
	// LlmCalls counts completed model calls (one per assistant message).
	LlmCalls int `json:"llm_calls"`
	// LlmDurationMs sums the provider-reported duration of those calls.
	LlmDurationMs int `json:"llm_duration_ms"`
	// ToolCalls counts the tool calls the turn recorded.
	ToolCalls int `json:"tool_calls"`
	// ToolDurationMs sums the measured duration of the tools that ran.
	ToolDurationMs int `json:"tool_duration_ms"`
	// PromptTokens, CompletionTokens and TotalTokens sum the provider-reported
	// usage. A provider that reports none leaves these at 0 rather than
	// estimating.
	PromptTokens     int `json:"prompt_tokens"`
	CompletionTokens int `json:"completion_tokens"`
	TotalTokens      int `json:"total_tokens"`
}

// sessionStatsFor aggregates one conversation's messages.
//
// It reads what is already in hand — the messages the handler just loaded — and
// adds one indexed aggregate query for the tool timing of turns that do not
// carry their own. A failure to read that aggregate is not a failure of the
// page: the counts, the tokens and the model durations are all still right, so
// the tool duration falls back to what the steps recorded rather than to a 500.
func (s *Server) sessionStatsFor(ctx context.Context, sessionID string, msgs []store.ChatMessage) sessionStats {
	var out sessionStats
	out.Messages = len(msgs)
	var toolMs int
	var sawToolMs bool
	for i := range msgs {
		m := &msgs[i]
		switch m.Role {
		case store.RoleUser:
			out.Turns++
		case store.RoleAssistant:
			// An assistant message that only carries reasoning or tool calls is
			// still one model call: the call happened, whatever it returned.
			out.LlmCalls++
			calls, ms, saw := transcriptTools(m)
			out.ToolCalls += calls
			if saw {
				toolMs += ms
				sawToolMs = true
			}
			// parseUsage is the same reader the transcript path uses, so a
			// message's usage is decoded exactly once, in one place.
			u := parseUsage(m.UsageJSON)
			out.LlmDurationMs += u.DurationMs
			out.PromptTokens += u.PromptTokens
			out.CompletionTokens += u.CompletionTokens
			out.TotalTokens += u.TotalTokens
		}
	}
	if sawToolMs {
		out.ToolDurationMs = toolMs
	}

	totals, err := s.store.QueryInvocationTotals(ctx, sessionID)
	if err != nil {
		s.logger.Warn("chat: read tool invocation totals failed, reporting what the steps recorded",
			zapString("session", sessionID), zapError(err))
		return out
	}
	// The audit log is the fallback, not the source: it is the only place an old
	// turn's timing exists at all (a turn stored before the steps kept their own),
	// and where the steps do have timing they are what the transcript shows, so
	// the two surfaces agree by construction.
	if !sawToolMs {
		out.ToolDurationMs = int(totals.DurationMs)
	}
	// A count that disagrees with the audit log means an invocation row was lost
	// or a call never ran. It is logged only when the audit log has rows at all:
	// a conversation whose turns predate the audit log is not a mismatch, it is
	// simply older than the log.
	if totals.Calls > 0 && totals.Calls != out.ToolCalls {
		s.logger.Debug("chat: tool call count differs from the audit log",
			zapString("session", sessionID),
			zap.Int("transcript", out.ToolCalls),
			zap.Int("audit", totals.Calls))
	}
	return out
}

// transcriptTools reads one assistant message's tool calls out of the transcript.
//
// The steps are the source because they are what the console renders: each one
// names the calls that iteration asked for, with the timing the call reported. A
// turn stored before steps were kept has none, and its calls live in the flat
// ToolCalls column — which is also what an older client reads — so that column is
// the fallback rather than a second thing to add up. Reading both and summing
// them would double every call of a current turn.
//
// `sawMs` reports whether any call in this message carried a duration: a message
// whose steps are all silent is "no timing recorded", which is not the same as
// 0ms and lets the caller fall back to the audit log.
func transcriptTools(m *store.ChatMessage) (calls, durationMs int, sawMs bool) {
	if m.Steps != "" {
		var steps []chat.Step
		if err := json.Unmarshal([]byte(m.Steps), &steps); err == nil {
			for _, step := range steps {
				for _, call := range step.Tools {
					calls++
					if call.DurationMs > 0 {
						durationMs += int(call.DurationMs)
						sawMs = true
					}
				}
			}
		}
	}
	if calls == 0 {
		// Either the turn kept no steps (an older row) or it kept steps that
		// called nothing: the flat list is the only other account of it.
		calls = countToolCalls(m.ToolCalls)
	}
	return calls, durationMs, sawMs
}

// countToolCalls counts the entries of one assistant message's tool_calls JSON.
// A row that does not parse counts as 0: the message is still a model call, and
// inventing a count from unreadable data would put a number on screen that
// nothing backs.
func countToolCalls(raw string) int {
	if raw == "" {
		return 0
	}
	var calls []json.RawMessage
	if err := json.Unmarshal([]byte(raw), &calls); err != nil {
		return 0
	}
	return len(calls)
}
